// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: api/gateway/v1/plan.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on CreatePlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CreatePlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreatePlanRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreatePlanRequestMultiError, or nil if none found.
func (m *CreatePlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreatePlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetPlan()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CreatePlanRequestValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CreatePlanRequestValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetPlan()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CreatePlanRequestValidationError{
				field:  "Plan",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CreatePlanRequestMultiError(errors)
	}

	return nil
}

// CreatePlanRequestMultiError is an error wrapping multiple validation errors
// returned by CreatePlanRequest.ValidateAll() if the designated constraints
// aren't met.
type CreatePlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreatePlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreatePlanRequestMultiError) AllErrors() []error { return m }

// CreatePlanRequestValidationError is the validation error returned by
// CreatePlanRequest.Validate if the designated constraints aren't met.
type CreatePlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreatePlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreatePlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreatePlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreatePlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreatePlanRequestValidationError) ErrorName() string {
	return "CreatePlanRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreatePlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreatePlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreatePlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreatePlanRequestValidationError{}

// Validate checks the field values on PlanModel with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PlanModel) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PlanModel with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PlanModelMultiError, or nil
// if none found.
func (m *PlanModel) ValidateAll() error {
	return m.validate(true)
}

func (m *PlanModel) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if l := utf8.RuneCountInString(m.GetName()); l < 1 || l > 20 {
		err := PlanModelValidationError{
			field:  "Name",
			reason: "value length must be between 1 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for PlanType

	// no validation rules for Cron

	// no validation rules for Enabled

	for idx, item := range m.GetDevices() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, PlanModelValidationError{
						field:  fmt.Sprintf("Devices[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, PlanModelValidationError{
						field:  fmt.Sprintf("Devices[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return PlanModelValidationError{
					field:  fmt.Sprintf("Devices[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	// no validation rules for NextExecTime

	for idx, item := range m.GetHistory() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, PlanModelValidationError{
						field:  fmt.Sprintf("History[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, PlanModelValidationError{
						field:  fmt.Sprintf("History[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return PlanModelValidationError{
					field:  fmt.Sprintf("History[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return PlanModelMultiError(errors)
	}

	return nil
}

// PlanModelMultiError is an error wrapping multiple validation errors returned
// by PlanModel.ValidateAll() if the designated constraints aren't met.
type PlanModelMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PlanModelMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PlanModelMultiError) AllErrors() []error { return m }

// PlanModelValidationError is the validation error returned by
// PlanModel.Validate if the designated constraints aren't met.
type PlanModelValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PlanModelValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PlanModelValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PlanModelValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PlanModelValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PlanModelValidationError) ErrorName() string { return "PlanModelValidationError" }

// Error satisfies the builtin error interface
func (e PlanModelValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPlanModel.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PlanModelValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PlanModelValidationError{}

// Validate checks the field values on CommandHistory with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CommandHistory) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CommandHistory with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CommandHistoryMultiError,
// or nil if none found.
func (m *CommandHistory) ValidateAll() error {
	return m.validate(true)
}

func (m *CommandHistory) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for CommandId

	// no validation rules for ExecTime

	// no validation rules for ExecResult

	if len(errors) > 0 {
		return CommandHistoryMultiError(errors)
	}

	return nil
}

// CommandHistoryMultiError is an error wrapping multiple validation errors
// returned by CommandHistory.ValidateAll() if the designated constraints
// aren't met.
type CommandHistoryMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CommandHistoryMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CommandHistoryMultiError) AllErrors() []error { return m }

// CommandHistoryValidationError is the validation error returned by
// CommandHistory.Validate if the designated constraints aren't met.
type CommandHistoryValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CommandHistoryValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CommandHistoryValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CommandHistoryValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CommandHistoryValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CommandHistoryValidationError) ErrorName() string { return "CommandHistoryValidationError" }

// Error satisfies the builtin error interface
func (e CommandHistoryValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCommandHistory.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CommandHistoryValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CommandHistoryValidationError{}

// Validate checks the field values on PlanDevice with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PlanDevice) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PlanDevice with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PlanDeviceMultiError, or
// nil if none found.
func (m *PlanDevice) ValidateAll() error {
	return m.validate(true)
}

func (m *PlanDevice) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for DeviceId

	if m.Switch1 != nil {

		if _, ok := _PlanDevice_Switch1_InLookup[m.GetSwitch1()]; !ok {
			err := PlanDeviceValidationError{
				field:  "Switch1",
				reason: "value must be in list [0 1]",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if m.Switch2 != nil {

		if _, ok := _PlanDevice_Switch2_InLookup[m.GetSwitch2()]; !ok {
			err := PlanDeviceValidationError{
				field:  "Switch2",
				reason: "value must be in list [0 1]",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if m.Switch3 != nil {

		if _, ok := _PlanDevice_Switch3_InLookup[m.GetSwitch3()]; !ok {
			err := PlanDeviceValidationError{
				field:  "Switch3",
				reason: "value must be in list [0 1]",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if m.SwitchMain != nil {

		if _, ok := _PlanDevice_SwitchMain_InLookup[m.GetSwitchMain()]; !ok {
			err := PlanDeviceValidationError{
				field:  "SwitchMain",
				reason: "value must be in list [0 1]",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if len(errors) > 0 {
		return PlanDeviceMultiError(errors)
	}

	return nil
}

// PlanDeviceMultiError is an error wrapping multiple validation errors
// returned by PlanDevice.ValidateAll() if the designated constraints aren't met.
type PlanDeviceMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PlanDeviceMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PlanDeviceMultiError) AllErrors() []error { return m }

// PlanDeviceValidationError is the validation error returned by
// PlanDevice.Validate if the designated constraints aren't met.
type PlanDeviceValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PlanDeviceValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PlanDeviceValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PlanDeviceValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PlanDeviceValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PlanDeviceValidationError) ErrorName() string { return "PlanDeviceValidationError" }

// Error satisfies the builtin error interface
func (e PlanDeviceValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPlanDevice.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PlanDeviceValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PlanDeviceValidationError{}

var _PlanDevice_Switch1_InLookup = map[int32]struct{}{
	0: {},
	1: {},
}

var _PlanDevice_Switch2_InLookup = map[int32]struct{}{
	0: {},
	1: {},
}

var _PlanDevice_Switch3_InLookup = map[int32]struct{}{
	0: {},
	1: {},
}

var _PlanDevice_SwitchMain_InLookup = map[int32]struct{}{
	0: {},
	1: {},
}

// Validate checks the field values on CreatePlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CreatePlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreatePlanReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreatePlanReplyMultiError, or nil if none found.
func (m *CreatePlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *CreatePlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return CreatePlanReplyMultiError(errors)
	}

	return nil
}

// CreatePlanReplyMultiError is an error wrapping multiple validation errors
// returned by CreatePlanReply.ValidateAll() if the designated constraints
// aren't met.
type CreatePlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreatePlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreatePlanReplyMultiError) AllErrors() []error { return m }

// CreatePlanReplyValidationError is the validation error returned by
// CreatePlanReply.Validate if the designated constraints aren't met.
type CreatePlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreatePlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreatePlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreatePlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreatePlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreatePlanReplyValidationError) ErrorName() string { return "CreatePlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e CreatePlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreatePlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreatePlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreatePlanReplyValidationError{}

// Validate checks the field values on UpdatePlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *UpdatePlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdatePlanRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdatePlanRequestMultiError, or nil if none found.
func (m *UpdatePlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdatePlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetPlan()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdatePlanRequestValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdatePlanRequestValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetPlan()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdatePlanRequestValidationError{
				field:  "Plan",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdatePlanRequestMultiError(errors)
	}

	return nil
}

// UpdatePlanRequestMultiError is an error wrapping multiple validation errors
// returned by UpdatePlanRequest.ValidateAll() if the designated constraints
// aren't met.
type UpdatePlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdatePlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdatePlanRequestMultiError) AllErrors() []error { return m }

// UpdatePlanRequestValidationError is the validation error returned by
// UpdatePlanRequest.Validate if the designated constraints aren't met.
type UpdatePlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdatePlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdatePlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdatePlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdatePlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdatePlanRequestValidationError) ErrorName() string {
	return "UpdatePlanRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdatePlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdatePlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdatePlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdatePlanRequestValidationError{}

// Validate checks the field values on UpdatePlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *UpdatePlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdatePlanReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdatePlanReplyMultiError, or nil if none found.
func (m *UpdatePlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdatePlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return UpdatePlanReplyMultiError(errors)
	}

	return nil
}

// UpdatePlanReplyMultiError is an error wrapping multiple validation errors
// returned by UpdatePlanReply.ValidateAll() if the designated constraints
// aren't met.
type UpdatePlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdatePlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdatePlanReplyMultiError) AllErrors() []error { return m }

// UpdatePlanReplyValidationError is the validation error returned by
// UpdatePlanReply.Validate if the designated constraints aren't met.
type UpdatePlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdatePlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdatePlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdatePlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdatePlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdatePlanReplyValidationError) ErrorName() string { return "UpdatePlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e UpdatePlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdatePlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdatePlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdatePlanReplyValidationError{}

// Validate checks the field values on DeletePlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *DeletePlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeletePlanRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeletePlanRequestMultiError, or nil if none found.
func (m *DeletePlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeletePlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return DeletePlanRequestMultiError(errors)
	}

	return nil
}

// DeletePlanRequestMultiError is an error wrapping multiple validation errors
// returned by DeletePlanRequest.ValidateAll() if the designated constraints
// aren't met.
type DeletePlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeletePlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeletePlanRequestMultiError) AllErrors() []error { return m }

// DeletePlanRequestValidationError is the validation error returned by
// DeletePlanRequest.Validate if the designated constraints aren't met.
type DeletePlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeletePlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeletePlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeletePlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeletePlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeletePlanRequestValidationError) ErrorName() string {
	return "DeletePlanRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeletePlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeletePlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeletePlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeletePlanRequestValidationError{}

// Validate checks the field values on DeletePlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *DeletePlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeletePlanReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeletePlanReplyMultiError, or nil if none found.
func (m *DeletePlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *DeletePlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return DeletePlanReplyMultiError(errors)
	}

	return nil
}

// DeletePlanReplyMultiError is an error wrapping multiple validation errors
// returned by DeletePlanReply.ValidateAll() if the designated constraints
// aren't met.
type DeletePlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeletePlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeletePlanReplyMultiError) AllErrors() []error { return m }

// DeletePlanReplyValidationError is the validation error returned by
// DeletePlanReply.Validate if the designated constraints aren't met.
type DeletePlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeletePlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeletePlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeletePlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeletePlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeletePlanReplyValidationError) ErrorName() string { return "DeletePlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e DeletePlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeletePlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeletePlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeletePlanReplyValidationError{}

// Validate checks the field values on GetPlanRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GetPlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetPlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GetPlanRequestMultiError,
// or nil if none found.
func (m *GetPlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetPlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetPlanRequestMultiError(errors)
	}

	return nil
}

// GetPlanRequestMultiError is an error wrapping multiple validation errors
// returned by GetPlanRequest.ValidateAll() if the designated constraints
// aren't met.
type GetPlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetPlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetPlanRequestMultiError) AllErrors() []error { return m }

// GetPlanRequestValidationError is the validation error returned by
// GetPlanRequest.Validate if the designated constraints aren't met.
type GetPlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetPlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetPlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetPlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetPlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetPlanRequestValidationError) ErrorName() string { return "GetPlanRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetPlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetPlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetPlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetPlanRequestValidationError{}

// Validate checks the field values on GetPlanReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GetPlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetPlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GetPlanReplyMultiError, or
// nil if none found.
func (m *GetPlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *GetPlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetPlan()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetPlanReplyValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetPlanReplyValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetPlan()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetPlanReplyValidationError{
				field:  "Plan",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetPlanReplyMultiError(errors)
	}

	return nil
}

// GetPlanReplyMultiError is an error wrapping multiple validation errors
// returned by GetPlanReply.ValidateAll() if the designated constraints aren't met.
type GetPlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetPlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetPlanReplyMultiError) AllErrors() []error { return m }

// GetPlanReplyValidationError is the validation error returned by
// GetPlanReply.Validate if the designated constraints aren't met.
type GetPlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetPlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetPlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetPlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetPlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetPlanReplyValidationError) ErrorName() string { return "GetPlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e GetPlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetPlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetPlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetPlanReplyValidationError{}

// Validate checks the field values on ListPlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListPlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListPlanRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListPlanRequestMultiError, or nil if none found.
func (m *ListPlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListPlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListPlanRequestMultiError(errors)
	}

	return nil
}

// ListPlanRequestMultiError is an error wrapping multiple validation errors
// returned by ListPlanRequest.ValidateAll() if the designated constraints
// aren't met.
type ListPlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListPlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListPlanRequestMultiError) AllErrors() []error { return m }

// ListPlanRequestValidationError is the validation error returned by
// ListPlanRequest.Validate if the designated constraints aren't met.
type ListPlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListPlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListPlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListPlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListPlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListPlanRequestValidationError) ErrorName() string { return "ListPlanRequestValidationError" }

// Error satisfies the builtin error interface
func (e ListPlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListPlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListPlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListPlanRequestValidationError{}

// Validate checks the field values on ListPlanReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ListPlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListPlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ListPlanReplyMultiError, or
// nil if none found.
func (m *ListPlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *ListPlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetPlans() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListPlanReplyValidationError{
						field:  fmt.Sprintf("Plans[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListPlanReplyValidationError{
						field:  fmt.Sprintf("Plans[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListPlanReplyValidationError{
					field:  fmt.Sprintf("Plans[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListPlanReplyMultiError(errors)
	}

	return nil
}

// ListPlanReplyMultiError is an error wrapping multiple validation errors
// returned by ListPlanReply.ValidateAll() if the designated constraints
// aren't met.
type ListPlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListPlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListPlanReplyMultiError) AllErrors() []error { return m }

// ListPlanReplyValidationError is the validation error returned by
// ListPlanReply.Validate if the designated constraints aren't met.
type ListPlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListPlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListPlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListPlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListPlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListPlanReplyValidationError) ErrorName() string { return "ListPlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e ListPlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListPlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListPlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListPlanReplyValidationError{}

// Validate checks the field values on SwitchPlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *SwitchPlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SwitchPlanRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SwitchPlanRequestMultiError, or nil if none found.
func (m *SwitchPlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SwitchPlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Enable

	if len(errors) > 0 {
		return SwitchPlanRequestMultiError(errors)
	}

	return nil
}

// SwitchPlanRequestMultiError is an error wrapping multiple validation errors
// returned by SwitchPlanRequest.ValidateAll() if the designated constraints
// aren't met.
type SwitchPlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SwitchPlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SwitchPlanRequestMultiError) AllErrors() []error { return m }

// SwitchPlanRequestValidationError is the validation error returned by
// SwitchPlanRequest.Validate if the designated constraints aren't met.
type SwitchPlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SwitchPlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SwitchPlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SwitchPlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SwitchPlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SwitchPlanRequestValidationError) ErrorName() string {
	return "SwitchPlanRequestValidationError"
}

// Error satisfies the builtin error interface
func (e SwitchPlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSwitchPlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SwitchPlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SwitchPlanRequestValidationError{}

// Validate checks the field values on SwitchPlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *SwitchPlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SwitchPlanReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SwitchPlanReplyMultiError, or nil if none found.
func (m *SwitchPlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *SwitchPlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetPlan()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SwitchPlanReplyValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SwitchPlanReplyValidationError{
					field:  "Plan",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetPlan()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SwitchPlanReplyValidationError{
				field:  "Plan",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return SwitchPlanReplyMultiError(errors)
	}

	return nil
}

// SwitchPlanReplyMultiError is an error wrapping multiple validation errors
// returned by SwitchPlanReply.ValidateAll() if the designated constraints
// aren't met.
type SwitchPlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SwitchPlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SwitchPlanReplyMultiError) AllErrors() []error { return m }

// SwitchPlanReplyValidationError is the validation error returned by
// SwitchPlanReply.Validate if the designated constraints aren't met.
type SwitchPlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SwitchPlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SwitchPlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SwitchPlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SwitchPlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SwitchPlanReplyValidationError) ErrorName() string { return "SwitchPlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e SwitchPlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSwitchPlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SwitchPlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SwitchPlanReplyValidationError{}

// Validate checks the field values on ExecPlanRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ExecPlanRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExecPlanRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExecPlanRequestMultiError, or nil if none found.
func (m *ExecPlanRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ExecPlanRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ExecPlanRequestMultiError(errors)
	}

	return nil
}

// ExecPlanRequestMultiError is an error wrapping multiple validation errors
// returned by ExecPlanRequest.ValidateAll() if the designated constraints
// aren't met.
type ExecPlanRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExecPlanRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExecPlanRequestMultiError) AllErrors() []error { return m }

// ExecPlanRequestValidationError is the validation error returned by
// ExecPlanRequest.Validate if the designated constraints aren't met.
type ExecPlanRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExecPlanRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExecPlanRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExecPlanRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExecPlanRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExecPlanRequestValidationError) ErrorName() string { return "ExecPlanRequestValidationError" }

// Error satisfies the builtin error interface
func (e ExecPlanRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExecPlanRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExecPlanRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExecPlanRequestValidationError{}

// Validate checks the field values on ExecPlanReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExecPlanReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExecPlanReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExecPlanReplyMultiError, or
// nil if none found.
func (m *ExecPlanReply) ValidateAll() error {
	return m.validate(true)
}

func (m *ExecPlanReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Message

	if len(errors) > 0 {
		return ExecPlanReplyMultiError(errors)
	}

	return nil
}

// ExecPlanReplyMultiError is an error wrapping multiple validation errors
// returned by ExecPlanReply.ValidateAll() if the designated constraints
// aren't met.
type ExecPlanReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExecPlanReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExecPlanReplyMultiError) AllErrors() []error { return m }

// ExecPlanReplyValidationError is the validation error returned by
// ExecPlanReply.Validate if the designated constraints aren't met.
type ExecPlanReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExecPlanReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExecPlanReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExecPlanReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExecPlanReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExecPlanReplyValidationError) ErrorName() string { return "ExecPlanReplyValidationError" }

// Error satisfies the builtin error interface
func (e ExecPlanReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExecPlanReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExecPlanReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExecPlanReplyValidationError{}
